felix/statusrep/status_file_reporter.go

```diff
@@ -0,0 +1,202 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statusrep
+
+import (
+	"context"
+	"errors"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/projectcalico/calico/felix/proto"
+
+	"github.com/sirupsen/logrus"
+)
+
+// EndpointStatusFileReporter writes a file to the FS
+// any time it sees an Endpoint go up in the dataplane.
+//
+//   - Currently only writes to a directory "policy", creating
+//     an entry for each workload, when each workload's
+//     policy is programmed for the first time.
+type EndpointStatusFileReporter struct {
+	endpointUpdatesC        <-chan interface{}
+	endpointStatusDirPrefix string
+
+	// Wraps and manages a real or mock wait.Backoff.
+	bom backoffManager
+}
+
+// Backoff wraps a timer-based-retry type which can be stepped.
+type Backoff interface {
+	Step() time.Duration
+}
+
+// backoffManager wraps and manages a real or mock wait.Backoff.
+type backoffManager struct {
+	Backoff
+	newBackoffFunc func() Backoff
+}
+
+// newBackoffManager creates a backoffManager which uses the
+// passed func to create backoffs.
+func newBackoffManager(newBackoffFunc func() Backoff) backoffManager {
+	return backoffManager{
+		Backoff:        newBackoffFunc(),
+		newBackoffFunc: newBackoffFunc,
+	}
+}
+
+// Reset the manager's backoff to its original state.
+func (bom *backoffManager) reset() {
+	bom.Backoff = bom.newBackoffFunc()
+}
+
+// Gets the manager's backoff.
+func (bom *backoffManager) getBackoff() Backoff {
+	return bom.Backoff
+}
+
+// FileReporterOption allows modification of a new EndpointStatusFileReporter.
+type FileReporterOption func(*EndpointStatusFileReporter)
+
+// NewEndpointStatusFileReporter creates a new EndpointStatusFileReporter.
+func NewEndpointStatusFileReporter(
+	endpointUpdatesC <-chan interface{},
+	statusDirPath string,
+	opts ...FileReporterOption,
+) *EndpointStatusFileReporter {
+
+	sr := &EndpointStatusFileReporter{
+		endpointUpdatesC:        make(<-chan interface{}),
+		endpointStatusDirPrefix: statusDirPath,
+
+		bom: newBackoffManager(newDefaultBackoff),
+	}
+
+	for _, o := range opts {
+		o(sr)
+	}
+
+	return sr
+}
+
+// WithNewBackoffFunc returns a FileReporterOption which alters the backoff
+// used by the reporter's backoff manager.
+func WithNewBackoffFunc(newBackoffFunc func() Backoff) FileReporterOption {
+	return func(fr *EndpointStatusFileReporter) {
+		fr.bom = newBackoffManager(newBackoffFunc)
+	}
+}
+
+// SyncForever blocks until ctx is cancelled.
+// Continuously pulls from endpoint-updates C,
+// and writes files to the status directory.
+func (fr *EndpointStatusFileReporter) SyncForever(ctx context.Context) {
+	recreatePolicyDir := true
+	for {
+		if recreatePolicyDir {
+			logrus.Debug("Creating/truncating policy status directory")
+			err := ensurePolicyStatusDir(fr.endpointStatusDirPrefix, true)
```

Comment:
I'd prefer if you synced the directory contents instead of deleting everything and starting over.  I'd use a DeltaTracker[string,bool], where the string is filename and the bool is true if the file should exist (on the Desired() side)/does exist(on the Dataplane() side).

Tracking it like that will give you a way to handle corner cases such as disk full since the DeltaTracker will keep track of the desired state so you can retry creating the file later.  It's also trivial to load the directory contents back up in to the Dataplane side with ReplaceAllIter.

Reply:
Of course, you'll need an "in sync" message to know when to delete old files.

Reply:
Would you be able to check what I've done with a SetDeltaTracker here? I may be misunderstanding the reason you recommended the base DeltaTracker, as opposed to SetDeltaTracker, or exactly what SetDeltaTracker is doing under the covers, in-which-case I'll adapt this to use the base DeltaTracker.